src/dataflow/src/sink/kafka.rs

```diff
@@ -626,6 +644,30 @@ impl KafkaSinkState {
     }
 
     async fn flush(&self) -> KafkaResult<()> {
+        loop {
+            self.flush_inner().await?;
+            if self.outstanding_send_count.load(Ordering::SeqCst) == 0
+                && self.retry_queue.lock().unwrap().len() == 0
+            {
+                break;
+            }
+            while let Some(msg) = { self.retry_queue.lock().unwrap().pop_front() } {
```

Comment:
can a thread be enqueuing at the same time as this is flushing? if so, this could never halt, we may need to have some max number of messages we try to flush at a time (unless we go with my above suggestion of always just re-enqueing onto rdkafka in my other review comment

Reply:
yeah I added the `{ }` to make it explicit but I'm happy to make it even more clear lest it be misunderstood when someone comes along later to read this!

Reply:
> can a thread be enqueuing at the same time as this is flushing? if so, this could never halt
yes -- and, in fact, that's the behavior we want! (and is consistent with the rest of the kafka sink's approach to errors).  If we can't send out these messages, we can't correctly make any other progress.

I'll try to see if I can think about how we might back off this type of errors like we do any other error type in the sink -- it seems tricker to track because they're flowing through kafka.  Do you have any ideas about that?